src/rest-server/src/models/v2/task.js

```diff
@@ -0,0 +1,125 @@
+} = require('@pai/models/v2/utils/frameworkConverter');
+
+const get = async (frameworkName, jobAttemptIndex, taskRoleName, taskIndex) => {
+  // get last task attempt by querying Framework / FrameworkHistory table
+  let attemptFramework;
+  const encodedFrameworkName = encodeName(frameworkName);
+
+  const framework = await databaseModel.Framework.findOne({
+    attributes: ['snapshot'],
+    where: { name: encodedFrameworkName },
+  });
+
+  if (framework) {
+    attemptFramework = JSON.parse(framework.snapshot);
+  } else {
+    throw createError(
+      'Not Found',
+      'NoJobError',
+      `Job ${frameworkName} is not found.`,
+    );
+  }
+
+  // when attemptIndex is not the last attempt, get the framework attempt from frameworkHistory table
+  if (jobAttemptIndex < attemptFramework.spec.retryPolicy.maxRetryCount) {
+    const historyFramework = await databaseModel.FrameworkHistory.findOne({
+      attributes: ['snapshot'],
+      where: {
+        frameworkName: encodedFrameworkName,
+        attemptIndex: jobAttemptIndex,
+      },
+    });
+
+    if (historyFramework) {
+      attemptFramework = JSON.parse(historyFramework.snapshot);
+    } else {
+      throw createError(
+        'Not Found',
+        'NoJobError',
+        `Job attempt not found in job ${frameworkName} with jobAttemptIndex ${jobAttemptIndex}.`,
+      );
+    }
+  } else if (
+    jobAttemptIndex > attemptFramework.spec.retryPolicy.maxRetryCount
+  ) {
+    throw createError(
+      'Not Found',
+      'NoJobError',
+      `Job attempt not found in job ${frameworkName} with jobAttemptIndex ${jobAttemptIndex}`,
+    );
+  }
+
+  // set task level info with `attemptFramework`
+  let taskStatus;
+  const taskRoleStatus = attemptFramework.status.attemptStatus.taskRoleStatuses.find(
+    (taskRoleStatus) => taskRoleStatus.name === taskRoleName,
+  );
+  if (taskRoleStatus) {
+    taskStatus = taskRoleStatus.taskStatuses.find(
+      (taskRoleStatus) => taskRoleStatus.index === taskIndex,
+    );
+  }
+  if (taskStatus === undefined) {
+    throw createError(
+      'Not Found',
+      'NoJobError',
+      `Task not found in Job ${frameworkName} with jobAttemptIndex ${jobAttemptIndex}, taskRoleName ${taskRoleName} and taskIndex ${taskIndex}.`,
+    );
+  }
+
+  // get task attempt history by querying TaskHistory table
+  const taskUid = taskStatus.instanceUID;
+  let taskHistories;
+  try {
+    taskHistories = await databaseModel.TaskHistory.findAll({
+      attributes: ['snapshot'],
+      where: { taskUid: taskUid },
+      order: [['attemptIndex', 'ASC']],
+    });
+  } catch (error) {
+    logger.error(`error when getting task from database: ${error.message}`);
+    throw error;
+  }
+
+  if (taskHistories) {
+    taskHistories = taskHistories.map((taskHistory) =>
+      JSON.parse(taskHistory.snapshot),
+    );
+  }
+
+  const taskDetail = await convertToTaskDetail(
+    attemptFramework,
+    taskRoleName,
+    taskStatus,
+    taskHistories,
```

Comment:
If role not found in job, why it returns null instead of not found?
See [https://int-bed/rest-server/api/v2/jobs/yqwang~taskhistory-elastic-2/attempts/0/taskRoles/taskrole/taskIndex/0/attempts](https://int-bed/rest-server/api/v2/jobs/yqwang~taskhistory-elastic-2/attempts/0/taskRoles/taskrole/taskIndex/0/attempts)